src/main/fc/fc_dispatch.c

```diff
@@ -62,6 +62,10 @@ void dispatchAdd(dispatchEntry_t *entry, int delayUs)
     dispatchEntry_t **p = &head;
     while (*p && cmp32((*p)->delayedUntil, delayedUntil) < 0)
         p = &(*p)->next;
-    entry->next = *p;
-    *p = entry;
+
+    /* Check for deadlock. Do not allow dispatching the same task twice */
```

Comment:
This test is not enough, it will fail if another entry has `delayedUntil` identical to duplicate entry ..

`delayedUntil` on enqueued entry is overwritten - timestamps won't be in monotonic order, causing all kinds of problems.

Reply:
OK, that is another problem. Two entries requesting the same delay. But will that not sort it self out, only with one of them delayed a bit more than requested? I can mot see that it could cause a deadlock, or?

Reply:
@ledvinap Ahh, now I see your concern, I think... Would it help to simply move one line?, like this:
```
 void dispatchAdd(dispatchEntry_t *entry, int delayUs)
 {
     uint32_t delayedUntil = micros() + delayUs;
-    entry->delayedUntil = delayedUntil;
     dispatchEntry_t **p = &head;
     while (*p && cmp32((*p)->delayedUntil, delayedUntil) < 0)
         p = &(*p)->next;
@@ -66,6 +65,7 @@ void dispatchAdd(dispatchEntry_t *entry, int delayUs)
     /* Check for deadlock. Do not allow dispatching the same task twice */
     if ( *p != entry) {
         entry->next = *p;
+        entry->delayedUntil = delayedUntil;
         *p = entry;
     }
 }
```

Reply:
@AndersHoglund : Problem is that you can't enqueue one entry twice.
Either caller ensures this (which may be preferable in this case), or this code has to check.

Moving `entry->delayedUntil` won't help, it may still still create circular list (only entry at insertion place is checked).

Either mark entry as 'in queue', or scan whole list (possibly using old entry->delayedUntil to optimize it a bit) to check for it ...

Reply:
@ledvinap Not sure I follow. But OK, I only check for consecutive callers being the same. Duplication and circular list may still occur if another caller comes inbetween.  Is this your concern? 
I did try and make some improvements on the caller side, to no avail at all. There are already some code for serialization, but in some case when the task is starved for CPU it does not seem to help. The only fix I found working was this change in the dispatcher. Tricky stuff this is.

Reply:
IMO the problem is that when entry is enqueued twice. 
It may create circular list (some entry is pointing to old version, next pointer is changed to point to part of list before this entry), leading to deadlock. 
Or it may break the linked list, orphanning some entries, possibly with nasty consequences (code registering in dispatcher won't be called ever again).

Only way to fix this is to prevent adding dispatch entry twice. Caller may handle this, but on dispatcher side there are few possibilities:

- mark entry as 'in queue' on add and clear this flag on dispatch. Check flag on Add

- scan whole linked list to check if entry is already in it.

Reply:
OK, got it. Did a test with an inQue element. Worked fine. Thank you @ledvinap .